vault-tool/src/Network/VaultTool/TOTP.hs

```diff
@@ -0,0 +1,252 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{- | A library for working with Vault's TOTP secrets engine
+   https://www.vaultproject.io/api-docs/secret/totp
+
+ Unless otherwise specified, all IO functions in this module may
+ potentially throw 'HttpException' or 'VaultException'
+-}
+module Network.VaultTool.TOTP (
+    GenerateKeyRequest (..),
+    HashAlgorithm (..),
+    DigitCount (..),
+    Skew (..),
+    GeneratedKey (..),
+    generateKey,
+
+    Key (..),
+    getKey,
+
+    KeyNames (..),
+    listKeys,
+
+    deleteKey,
+
+    Code (..),
+    generateCode,
+
+    CodeStatus (..),
+    validateCode,
+) where
+
+import Data.Aeson (
+    FromJSON (..),
+    ToJSON (..),
+    Value (..),
+    (.:),
+    withObject,
+ )
+import Data.Aeson.Utils (DataWrapper (..), (.=!), (.=?), object)
+import Data.Bool (bool)
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import Network.VaultTool.Internal (
+    newDeleteRequest,
+    newGetRequest,
+    newListRequest,
+    newPostRequest,
+    runVaultRequestAuthenticated,
+    runVaultRequestAuthenticated_,
+    withStatusCodes,
+ )
+import Network.VaultTool.Types (
+    Authenticated,
+    VaultConnection,
+    VaultMountedPath (..),
+ )
+
+-- | The data needed for vault to generate a TOTP Key
+-- <https://www.vaultproject.io/api-docs/secret/totp#parameters>
+data GenerateKeyRequest = GenerateKeyRequest
```

Comment:
How do you feel about defining and exporting a smart constructor which sets all the `Maybe` fields to `Nothing`?  That would make it easy to compose values like
```haskell
(generateKeyRequest keyName issuer accountName) { gkrPeriod = Just 120 }
```

Reply:
I considered this but didn't end up going that route because I got hung up on whether or not I should have
```
mkGenerateKeyRequest :: Text -> Text -> Text -> GenerateKeyRequest
```
or
```
mkGenerateKeyRequest :: KeyName -> Issuer -> AccountName -> GenerateKeyRequest
```

The order of the args is ambiguous in the former and the latter requires us to add newtype wrappers. If we add those newtype wrappers do we add wrappers for all the other fields in `GenerateKeyRequest` as well?

Reply:
One compromise is to use type aliases e.g. `type KeyName = Text`.  It doesn't help if somebody gets confused and puts the args in the wrong order, but makes the signature more clear.

Reply:
What are your thoughts on how handle setting the other fields? Exporting the record fields so we can do record updates vs hiding the fields and providing setter functions?
```
(generateKeyRequest keyName issuer accountName) { gkrPeriod = Just 120 }
```
vs
```
withPeriod 120 (generateKeyRequest keyName issuer accountName)
```
I don't see a reason why the fields need to be exported but also I don't know that there's any downsides. Typically I've gone the setter function route but that's just personal preference and I'm wondering if there's a compelling reason to go one route over the other.

Reply:
In this case, `GenerateKeyRequest` models something external, specifically the parameters for generating a key as defined by Vault.  This is the safest situation in which to export the entire record, since internal implementation changes are not going to affect the type.

The main advantages of exposing the constructor and fields is that it is the most straightforward.  All the available fields are right there in the definition.  

Exporting a bunch of setters has an abstraction benefit, but spreads the structure of the value out over a whole set of functions and types.  I consider this to be an important cost, and not worth it in this situation.

So IMO exposing the record ctor and fields is best.

Reply:
0854456b93adb152efa7cd305d16354f88138545